### ISSUE TITLE
Enhancement: Only show focus outlines on keyboard interaction

### ISSUE THREAD

**Issue by patrickarlt** [Esri/calcite-components#1359]
### Description

Most Calcite components show focus rings when clicked such as in the following case with `<calcite-button>`:

![image](https://user-images.githubusercontent.com/378557/101210509-7090db80-362a-11eb-9cd9-d3450a41846c.png)

These focus states are almost entirely unnecessary for mouse interactions and are a common complaint of people I see using Calcite Components. With Edge using Chromium now almost all browsers except Safari support `:focus-within` which only triggers on keyboard interactions so you can do this https://codepen.io/patrickarlt/pen/poEgXmv?editors=0100

```css
/* enable a default focus state for legacy browsers */
button:focus {
  outline: 2px solid hotpink;
  outline-offset: 2px;
}

/* 
 * disable the focus if it is has focus but NOT focus-visible
 * non supporting browsers will ignore this rule because they
 * don't understand focus-visible
 */
button:focus:not(:focus-visible) {
  outline: none;
}

/* now add a focus + focus visible state for keyboard focus only */
button:focus:focus-visible {
  outline: 2px solid hotpink;
  outline-offset: 2px;
}
```

Although Safari doesn't support `focus-within` for whatever reason it also doesn't seem to apply focus styles on pointer interactions by default.

The CodePen demo also shows that elements given programmatic focus still show styles with `focus-within`.

### Acceptance Criteria 

Only show focus styles on programmatic focus for explicit keyboard interaction

### Relevant Info 

* CanIUse: https://caniuse.com/css-focus-visible
* CodePen Demo: https://codepen.io/patrickarlt/pen/poEgXmv
* Google Dev article: https://blog.chromium.org/2020/09/giving-users-and-developers-more.html

#### Which Component

All components with `:focus` styles.

#### Example Use Case

![image](https://user-images.githubusercontent.com/378557/101210509-7090db80-362a-11eb-9cd9-d3450a41846c.png)

**Comment by patrickarlt** [comment]
I thought IE 11 supported `@supports` but it doesn't https://caniuse.com/?search=%40supports. You can still do this like:

```css

button:focus {
  outline: 2px solid hotpink;
  outline-offset: 2px;
}

@supports (:focus-within) {
  button:focus {
    outline: none;
  }
  button:focus-within {
    outline: 2px solid hotpink;
    outline-offset: 2px;
  }
}
```

**Comment by patrickarlt** [comment]
This seems to be the proper method for this that works everywhere. Chrome doesn't seem to support `@supports(:focus-within)`

```css
/* enable a default focus state for legacy browsers */
button:focus {
  outline: 2px solid hotpink;
  outline-offset: 2px;
}

/* 
 * disable the focus if it is has focus but NOT focus-visible
 * non supporting browsers will ignore this rule because they
 * don't understand focus-visible
 */
button:focus:not(:focus-visible) {
  outline: none;
}

/* 
 * now add a focus + focus visible state for keyboard focus only
 * this cannot be combined with the first :focus rule for some reason.
 */
button:focus:focus-visible {
  outline: 2px solid hotpink;
  outline-offset: 2px;
}
```

**Comment by macandcheese** [comment]
I think aesthetically this is desirable but we need to make sure there aren't accessibility implications from it - I think the concern being some assistive devices use simulated mouse presses as a method of interaction. 

An argument FOR this would be that Safari seems to disable these focus rings out of the box (using the example of Discover.com - in Chrome rings apply on click, in Safari they do not).

**Comment by julio8a** [comment]
The component library is getting an audit by level access, will bring this up.

**Comment by geospatialem** [comment]
Since the issue's creation, much more support has been added across browsers. 

As @patrickarlt mentioned IE seems to be the only holdout, but there are workarounds. Plus IE is no longer maintained by Microsoft, and `:focus-visible` is supported by Edge.

As long as it is implemented as Patrick mentions above, this is a great enhancement across the components for those who navigate via mouse, while not taking away functionality for keyboard users. 

We will need to ensure we're not using `:focus { outline: 0; }` 🚫, which equates to **total keyboard inaccessibility**.

**Comment by macandcheese** [comment]
I know in the past when this has come up there was concern over how it could impact low-sight mouse users. We should validate that concern before proceeding.

**Comment by geospatialem** [comment]
Going to add this topic to our upcoming Core meeting to discuss further.

**Comment by geospatialem** [comment]
Per the team discussion:
- Mouse users can benefit from the focus outline, in particular those with low sight.
- There are instances where block elements could be mitigated further, and will be investigated in 4633

With this course of action, we are closing this issue in favor of #4633.
